stablecoin-v3/src/lib.rs

```diff
@@ -278,6 +299,155 @@ pub trait StablecoinV3:
+
+        let stablecoin_token_id = self.stablecoin().get_token_id();
+        let base_collateral_token_id = self.base_collateral_token_id().get();
+
+        let (stablecoin_rewards, collateral_rewards) =
+            self.calculate_fee_rewards(&cp_token_id, payment_nonce, &payment_amount);
+        self.reward_reserve(&stablecoin_token_id)
+            .update(|x| *x -= &stablecoin_rewards);
+        self.reward_reserve(&base_collateral_token_id)
+            .update(|x| *x -= &collateral_rewards);
+        self.burn_cp_tokens(&cp_token_id, payment_nonce, &payment_amount);
+
+        let current_epoch = self.blockchain().get_block_epoch();
+
+        let virtual_position = CpTokenAttributes {
+            stablecoin_reward_per_share: self.reward_per_share(&stablecoin_token_id).get(),
+            collateral_reward_per_share: self.reward_per_share(&base_collateral_token_id).get(),
+            entering_epoch: current_epoch,
+        };
+
+        let user_cp_payment = self.mint_cp_tokens(cp_token_id, payment_amount, &virtual_position);
+
+        let caller = self.blockchain().get_caller();
+
+        let mut payments = ManagedVec::new();
+
+        payments.push(EsdtTokenPayment::new(
+            stablecoin_token_id,
+            0,
+            stablecoin_rewards,
+        ));
+        payments.push(EsdtTokenPayment::new(
+            base_collateral_token_id,
+            0,
+            collateral_rewards,
+        ));
+        payments.push(user_cp_payment);
+
+        self.send().direct_multi(&caller, &payments);
+    }
+
+    #[view(calculateFeeRewards)]
+    fn calculate_fee_rewards(
+        &self,
+        cp_token_id: &TokenIdentifier,
+        nonce: u64,
+        amount: &BigUint,
+    ) -> (BigUint, BigUint) {
+        let cp_token_attributes =
+            self.get_cp_token_attributes::<CpTokenAttributes<Self::Api>>(cp_token_id, nonce);
+        let stablecoin_rps = self
+            .reward_per_share(&self.stablecoin().get_token_id())
+            .get();
+        let collateral_rps = self
+            .reward_per_share(&self.base_collateral_token_id().get())
+            .get();
+
+        let stablecoin_rewards =
+            if &stablecoin_rps > &cp_token_attributes.stablecoin_reward_per_share {
+                let rps_diff = &stablecoin_rps - &cp_token_attributes.stablecoin_reward_per_share;
+                let div_safety = self.division_safety_constant().get();
+
+                amount * &rps_diff / div_safety
+            } else {
+                BigUint::zero()
+            };
+
+        let collateral_rewards =
+            if &collateral_rps > &cp_token_attributes.collateral_reward_per_share {
+                let rps_diff = &collateral_rps - &cp_token_attributes.collateral_reward_per_share;
+                let div_safety = self.division_safety_constant().get();
+
+                amount * &rps_diff / div_safety
+            } else {
+                BigUint::zero()
+            };
```

Comment:
Duplicated logic.
Suggestion - move the logic to a new method `compute_rewards` and call it twice with something like:
```
let stablecoin_rewards = self.compute_rewards(
    self.stablecoin().get_token_id(),
    &cp_token_attributes.stablecoin_reward_per_share
);
let collateral_rewards = self.compute_rewards(
    self.base_collateral_token_id().get(),
    &cp_token_attributes.collateral_reward_per_share
);
```